### DISCUSSION TITLE
how to printout node number with std::cout

### DISCUSSION THREAD

**Discussion by styyokuda** [idaholab/moose#17803] [Q&A General] [answered]
Dear MOOSE Team,

I am working on debugging by printing variable "flux_value" using std::cout as:
std::cout << " " << _q_point[_qp] 
              << " flux = " << _flux_value[_qp] 
              << " dt =" << _dt << " t =" << _t
              << std::endl;
I am also trying to print the node number using "_current_node" but it is not working.
Please help.
Thomas

**Comment by GiudGiud** [comment]
The '<<' operator probably has not been defined for the type of _current_node (libmesh Node). You can print its attributes or you can use the already defined routines.

See [libmesh Node](https://libmesh.github.io/doxygen/classlibMesh_1_1Node.html)

Note that if you used a debugger instead I think you could just do `p _current_node` and get everything about it. 

Guillaume

**Reply by styyokuda** [reply]
Thank you GiudGiud.
No. I am not using a debugger. 
I am still new to MOOSE.
Could you please explain more about "You can print its attributes or you can use the already defined routines?"
It would be appreciated if you could give me an example(s).
Thomas

**Reply by GiudGiud** [reply]
like you can do std::cout << _current_node.get_info() or .get_id()

**Reply by styyokuda** [reply]
I am sorry Guillaume but
std::cout << _current_node.get_info() 
or
std::cout << _current_node.get_id()
are not working for me. :(
Thomas

**Reply by loganharbour** [reply]
Why are you trying to get the node ID?

Also, it's a pointer:
```
std::cout << _current_node->id() << std::endl;
```

**Comment by styyokuda** [comment]
Thank you Logan,
By some reason, "std::cout << _current_node->id() << std::endl;" is not
working for me. :(
In my program, coefficients are calculated and I am trying to make sure the
coefficients values are correct by printing them with the node number. If I
don't use "std::cout," what do you recommend?
Please let me know.
Thomas

**Reply by GiudGiud** [reply]
I believe it's get_id() you should be using. <- not true, see below

I recommend you use a debugger. They are incredibly powerful and can be used with any software. You will find tutorials online, and there is more information for using them in MOOSE [there](https://mooseframework.inl.gov/application_development/debugging.html)

**Reply by hugary1995** [reply]
No, `std::cout << _current_node->id() << std::endl;` should work. Please let us know what do you mean by "is not working for me", then we can help.

**Comment by styyokuda** [comment, marked as answer]
Thank you very much, Gary.
The compiler error is: *use of undeclared identifier '_current_node';**did
you mean '_current_elem'?*
std::cout << _current_elem->id() << std::endl; is working but this printout
is for element number, isn't it?
My MacBOOK Pro is still running on macOS Mojave and I have not updated
MOOSE for more than a year
because I believe a higher macOS operating system is required for the
update.
Do you think this is the reason?

It is not exactly debugging. Because I wrote C++ objects to calculate a
number of coefficients and boundary fluxes (for BC settings)
for a heat transfer and concentration diffusion model.
I thought I could print coefficients and fluxes along with temperatures and
concentrations on nodes on the screen with std::cout.
Please let me know if any other way is possible.
I would like to thank you for your time and help.
Thomas

**Reply by GiudGiud** [reply]
Ok _current_node is not set for the class of kernel that you are using. You should work off the _current_elem and _qp index to retrieve the node if you need it. 
See [libmesh Elem doxygen](https://libmesh.github.io/doxygen/classlibMesh_1_1Elem.html#ab0a3d359585810dbf06529da435382bf) for all the routines you can use. 

You dont need to update your macOS to update moose.

**Reply by styyokuda** [reply]
Thank you GiudGiud for your time and effort.
Because your suggestion seems to be too much work for me, I will look for other way to do that.
Thomas
